mltensor-explainer.md

```diff
@@ -0,0 +1,423 @@
+
+### Support passing an `MLTensor` as a `constant()`
+
+There may be an appetite for a mechanism to stream constant weights to the ML context without needing to go through an `ArrayBuffer`, but whether that mechanism should use the `MLTensor` interface is not yet clear. It's not a natural fit because an `MLTensor` is scoped to an `MLContext`, whereas a constant `MLOperand` is scoped to a single `MLGraphBuilder` (and subsequently the `MLGraph` it builds), since sharing constant data across graphs is [not reliably supported on some platforms](https://github.com/webmachinelearning/webnn/issues/614#issuecomment-2021581363).
+
+### Allow mapping an `MLTensor` to script
+
+Why doesn't `MLTensor` have a [`mapAsync()`](https://www.w3.org/TR/webgpu/#dom-gpubuffer-mapasync) method, as a `GPUBuffer` does?
+
+`MLTensor`s are used as inputs and outputs to machine learning models. In practice, we expect the size of the model inputs and outputs to be dwarfed by the size of the model weights, which are uploaded via the `MLGraphBuilder.constant()` method. We may re-evaluate this stance in the future if we discover that reading and writing data to an `MLTensor` is a bottleneck, though even in that case we may prefer to explore a solution which bypasses `ArrayBuffer`s altogether, as discussed [above](#support-passing-an-mltensor-as-a-constant).
+
+### Hash input buffer contents
+
+One approach to solve the [buffer reuse](#buffer-reuse) case is for the WebNN implementation to silently avoid making redundant data copies if the same buffer contents are repeatedly passed as inputs. This may be achieved by hashing the contents of each input. This approach has downsides, such as [managing the extra copies in the hash map](#resource-management) and that hashing the buffer contents may be expensive as it requires reading the entire input. This approach also does not address the other use cases.
+
+## References & Acknowledgements
+
+Many thanks for valuable feedback and advice from:
+
+- Bryan Bernhart
+- Joshua Bell
+- Mike Wyrzykowski
+- Ningxin Hu
+- Phillis Tang
+- Rafael Cintron
+- Reilly Grant
+- Zoltan Kis
+
+---
+
+## Appendix
+
+### Tentative IDL
+
+```javascript
+typedef [EnforceRange] unsigned long MLTensorUsageFlags;
+
+namespace MLTensorUsage {
+    const MLFlagsConstant READ_FROM      = 0x0001;
+    const MLFlagsConstant WRITE_TO       = 0x0002;
+    const MLFlagsConstant WEBGPU_INTEROP = 0x0004;
+};
+
+dictionary MLTensorDescriptor : MLOperandDescriptor {
+  required MLTensorUsageFlags usage;
+};
+
+typedef record<DOMString, MLTensor> MLNamedTensors;
+
+interface MLTensor {
+  readonly attribute MLOperandDataType dataType;
+  readonly attribute FrozenArray<unsigned long> shape;
+  readonly attribute unsigned long MLTensorUsageFlags usage;
+
+  void destroy();
+};
+
+partial interface MLContext {
+  Promise<MLTensor> createTensor(MLTensorDescriptor descriptor);
+
+  void writeBuffer(MLTensor dstTensor, [AllowShared] ArrayBuffer srcData);
```

Comment:
nit: writeTensor since this isn't a buffer like `GPUBuffer` but a tensor type.

Reply:
🤔 Conversely for `readBuffer`, we're not reading a GPU buffer anymore. We're reading a tensor (and amusingly, `readBuffer` actually *writes* into a buffer, an ArrayBuffer).

Reply:
Updated to `writeTensor()` and `readTensor()` for now (in line with https://crrev.com/c/5842431)